### Info2A-PSIx-PSI 2022-23-TD06_knn_CR

* énoncé téléchargeable : <a href="https://github.com/lbaille20/Info2A_2022-23/raw/main/Info2A-PSIx/TD06/Info2A-PSIx-PSI_TD06_2022-23_knn%20EN_v2.pdf">ici</a>

* ce notebook  : **0ba6-1241323**

Il y a en annexe de ce notebook trois fichiers de données, que l'on peut lister en exécutant les deux cellules ci-dessous.

In [1]:
import os

In [2]:
os.listdir()

['iris_data_ori.txt', 'iris.csv', 'iris_data_description_ori.txt']

Ces trois fichiers peuvent être téléchargés en cliquant sur le menu Capytale > Fichiers annexes > Attachés à l'activité > nom du fichier > Télécharger :

<img src = "https://github.com/lbaille20/Info2A_2022-23/raw/main/T%C3%A9l%C3%A9chargement%20fichiers%20joints.jpg"/>

On peut aussi télécharger ces fichiers en exécutant chacune des trois cellules suivantes :

* téléchargement du fichier de données

In [3]:
import basthon
basthon.download('iris.csv')

* fichiers de données originaux : http://archive.ics.uci.edu/ml/datasets/Iris

In [4]:
import basthon
basthon.download('iris_data_description_ori.txt')

In [5]:
import basthon
basthon.download('iris_data_ori.txt')

**Q1.**

La structure du fichier **iris.csv** est la suivante :

<img src="https://github.com/lbaille20/Info2A_2022-23/raw/main/Info2A-PSIx/TD06/iris-csv.png"/>

On peut donc utilement :
- ouvrir l'accès au fichier en récupérant l'objet-fichier (ou descripteur de fichier) `f` ;  
- lire la première ligne du fichier (`f.readline()`) et la stocker dans une variable `entetes` ;
- lire ensuite toutes les lignes suivantes en les stockant dans une liste (f.readlines()) `donnees` ;
- enfin fermer l'accès au fichier (`f.close()`).

In [6]:
f = open("iris.csv")
entetes = f.readline()
donnees = f.readlines()
f.close()

On a ainsi récupéré la ligne d'entêtes :

In [7]:
entetes

'petal_length,petal_width,species\n'

ainsi qu'une liste de 150 chaînes de caractères, correspondant chacune à la description d'une fleur de l'échantillon :

In [8]:
len(donnees)

150

On peut examiner les quatre premiers éléments de cette liste, correspondant aux quatre premières lignes de données du fichier, dont la capture d'écran ci-dessus a dévoilé le contenu :

In [9]:
donnees[:4]

['1.4,0.2,0\n', '1.4,0.2,0\n', '1.3,0.2,0\n', '1.5,0.2,0\n']

Pour chacune de ces lignes, par exemple ici la première :

In [10]:
ligne1 = donnees[0]
ligne1

'1.4,0.2,0\n'

on doit retirer le symbole de fin de ligne :

In [11]:
ligne1a = ligne1.strip() ## une nouvelle chaîne est créée par laquelle on remplace l'ancienne
ligne1a

'1.4,0.2,0'

puis en séparer les éléments en découpant la chaîne aux endroits marqués par le séparateur `,` et obtenir une liste de trois chaînes de caractères :

In [12]:
ligne1b = ligne1a.split(',') ## une liste de trois chaînes est créée par laquelle on remplace la chaîne traitée
ligne1b

['1.4', '0.2', '0']

En vue de l'utilisation ultérieure des données, on doit convertir les deux premiers éléments au format de flottants (ces valeurs, longueur et largeur des pétales, seront utilisés pour calculer la distance entre deux fleurs) :

In [13]:
longueur1, largeur1, espece1 = float(ligne1b[0]), float(ligne1b[1]), int(ligne1b[2])
longueur1, largeur1, espece1

(1.4, 0.2, 0)

La conversion du numéro désignant l'espèce au type `int` a été effectuée (demandée par l'énoncé), mais n'a pas de caractère de nécessité.

* traitement de l'ensemble des données

En suivant la démarche précédente appliquée au traitement de la première ligne de données, on peut traiter tout l'échantillon comme suit (où l'on a essayé de minimiser le nombre de variables utilisées sans nuire à la lisibilité du code) :

In [14]:
data = []

for ligne in donnees :
    ligne = ligne.strip("\n")
    ligne = ligne.split(",")
    data.append(ligne)

for i in range(len(data)):
    x, y = [float(data[i][j]) for j in range(2)]
    espece = int(data[i][2])
    data[i] = (x, y, espece) ## on remplace data[i] par le tuple demandé

Les données sont maintenant disponibles au bon format :

In [15]:
data

[(1.4, 0.2, 0), (1.4, 0.2, 0), (1.3, 0.2, 0), (1.5, 0.2, 0), (1.4, 0.2, 0), (1.7, 0.4, 0), (1.4, 0.3, 0), (1.5, 0.2, 0), (1.4, 0.2, 0), (1.5, 0.1, 0), (1.5, 0.2, 0), (1.6, 0.2, 0), (1.4, 0.1, 0), (1.1, 0.1, 0), (1.2, 0.2, 0), (1.5, 0.4, 0), (1.3, 0.4, 0), (1.4, 0.3, 0), (1.7, 0.3, 0), (1.5, 0.3, 0), (1.7, 0.2, 0), (1.5, 0.4, 0), (1.0, 0.2, 0), (1.7, 0.5, 0), (1.9, 0.2, 0), (1.6, 0.2, 0), (1.6, 0.4, 0), (1.5, 0.2, 0), (1.4, 0.2, 0), (1.6, 0.2, 0), (1.6, 0.2, 0), (1.5, 0.4, 0), (1.5, 0.1, 0), (1.4, 0.2, 0), (1.5, 0.1, 0), (1.2, 0.2, 0), (1.3, 0.2, 0), (1.5, 0.1, 0), (1.3, 0.2, 0), (1.5, 0.2, 0), (1.3, 0.3, 0), (1.3, 0.3, 0), (1.3, 0.2, 0), (1.6, 0.6, 0), (1.9, 0.4, 0), (1.4, 0.3, 0), (1.6, 0.2, 0), (1.4, 0.2, 0), (1.5, 0.2, 0), (1.4, 0.2, 0), (4.7, 1.4, 1), (4.5, 1.5, 1), (4.9, 1.5, 1), (4.0, 1.3, 1), (4.6, 1.5, 1), (4.5, 1.3, 1), (4.7, 1.6, 1), (3.3, 1.0, 1), (4.6, 1.3, 1), (3.9, 1.4, 1), (3.5, 1.0, 1), (4.2, 1.5, 1), (4.0, 1.0, 1), (4.7, 1.4, 1), (3.6, 1.3, 1), (4.4, 1.4, 1), (4.5, 1.5

Une version plus brève du traitement :

In [16]:
data1 = [ligne.strip().split(',') for ligne in donnees]
data1

[['1.4', '0.2', '0'], ['1.4', '0.2', '0'], ['1.3', '0.2', '0'], ['1.5', '0.2', '0'], ['1.4', '0.2', '0'], ['1.7', '0.4', '0'], ['1.4', '0.3', '0'], ['1.5', '0.2', '0'], ['1.4', '0.2', '0'], ['1.5', '0.1', '0'], ['1.5', '0.2', '0'], ['1.6', '0.2', '0'], ['1.4', '0.1', '0'], ['1.1', '0.1', '0'], ['1.2', '0.2', '0'], ['1.5', '0.4', '0'], ['1.3', '0.4', '0'], ['1.4', '0.3', '0'], ['1.7', '0.3', '0'], ['1.5', '0.3', '0'], ['1.7', '0.2', '0'], ['1.5', '0.4', '0'], ['1.0', '0.2', '0'], ['1.7', '0.5', '0'], ['1.9', '0.2', '0'], ['1.6', '0.2', '0'], ['1.6', '0.4', '0'], ['1.5', '0.2', '0'], ['1.4', '0.2', '0'], ['1.6', '0.2', '0'], ['1.6', '0.2', '0'], ['1.5', '0.4', '0'], ['1.5', '0.1', '0'], ['1.4', '0.2', '0'], ['1.5', '0.1', '0'], ['1.2', '0.2', '0'], ['1.3', '0.2', '0'], ['1.5', '0.1', '0'], ['1.3', '0.2', '0'], ['1.5', '0.2', '0'], ['1.3', '0.3', '0'], ['1.3', '0.3', '0'], ['1.3', '0.2', '0'], ['1.6', '0.6', '0'], ['1.9', '0.4', '0'], ['1.4', '0.3', '0'], ['1.6', '0.2', '0'], ['1.4', '0.2

et ensuite :

In [17]:
data = [(float(t[0]), float(t[1]), int(t[2])) for t in data1]
data

[(1.4, 0.2, 0), (1.4, 0.2, 0), (1.3, 0.2, 0), (1.5, 0.2, 0), (1.4, 0.2, 0), (1.7, 0.4, 0), (1.4, 0.3, 0), (1.5, 0.2, 0), (1.4, 0.2, 0), (1.5, 0.1, 0), (1.5, 0.2, 0), (1.6, 0.2, 0), (1.4, 0.1, 0), (1.1, 0.1, 0), (1.2, 0.2, 0), (1.5, 0.4, 0), (1.3, 0.4, 0), (1.4, 0.3, 0), (1.7, 0.3, 0), (1.5, 0.3, 0), (1.7, 0.2, 0), (1.5, 0.4, 0), (1.0, 0.2, 0), (1.7, 0.5, 0), (1.9, 0.2, 0), (1.6, 0.2, 0), (1.6, 0.4, 0), (1.5, 0.2, 0), (1.4, 0.2, 0), (1.6, 0.2, 0), (1.6, 0.2, 0), (1.5, 0.4, 0), (1.5, 0.1, 0), (1.4, 0.2, 0), (1.5, 0.1, 0), (1.2, 0.2, 0), (1.3, 0.2, 0), (1.5, 0.1, 0), (1.3, 0.2, 0), (1.5, 0.2, 0), (1.3, 0.3, 0), (1.3, 0.3, 0), (1.3, 0.2, 0), (1.6, 0.6, 0), (1.9, 0.4, 0), (1.4, 0.3, 0), (1.6, 0.2, 0), (1.4, 0.2, 0), (1.5, 0.2, 0), (1.4, 0.2, 0), (4.7, 1.4, 1), (4.5, 1.5, 1), (4.9, 1.5, 1), (4.0, 1.3, 1), (4.6, 1.5, 1), (4.5, 1.3, 1), (4.7, 1.6, 1), (3.3, 1.0, 1), (4.6, 1.3, 1), (3.9, 1.4, 1), (3.5, 1.0, 1), (4.2, 1.5, 1), (4.0, 1.0, 1), (4.7, 1.4, 1), (3.6, 1.3, 1), (4.4, 1.4, 1), (4.5, 1.5

**Q2.**

* nombre de fleurs décrites dans le fichier de données (c'est aussi la longueur de la liste `donnees`) :

In [18]:
len(data)

150

* dénombrement par espèce

En utilisant que les espèces sont repérées dans la liste de liste `data` par les entiers 0, 1 et 2, on peut dénombrer les différentes espèces de la façon suivante :

In [19]:
nbr_especes = [0,0,0]
for el in data :
    nbr_especes[el[2]] += 1

On obtient :

In [20]:
nbr_especes

[50, 50, 50]

On aurait pu aussi utiliser un dictionnaire pour collecter les résultats :

In [21]:
Dcpt_especes = {0: 0, 1: 0, 2: 0}
for elt in data :
    Dcpt_especes[elt[2]] += 1

In [22]:
Dcpt_especes

{0: 50, 1: 50, 2: 50}

**Q3.**

Pour obtenir rapidement la représentation graphique demandée, le code suivant convient, dans lequel, pour chaque fleur de l'échantillon, on place un point, affecté de la couleur demandée :

In [23]:
import matplotlib.pyplot as plt

fig, ax = plt.subplots() # nécessaire pour des affichages corrects dans Capytale

for triplet in data:
    longueur, largeur, espece = triplet
    if espece == 0:
        plt.plot([longueur], [largeur], 'ro') ## r pour red (iris setosa), o pour marquer le point par une "bille"
    elif espece == 1:
        plt.plot([longueur], [largeur], 'bo') ## b pour blue (iris vigininica), o pour marquer le point par une "bille"
    else:
        plt.plot([longueur], [largeur], 'go') ## g pour green (iris versicolor), o pour marquer le point par une "bille"
        
plt.show()

***On propose ci-dessous de produire la représentation graphique de façon plus structurée :***

On va construire trois listes, stockées dans un dictionnaire, `Dcoords`, dont les clés sont les noms des espèces.  
Chacune de ces trois listes sera une liste des couples (x, y) = (longueur, largeur) pour chaque fleur de l'espèce considérée.  

On pourra utilement se donner un dictionnaire dont les clés sont les numéros des espèces, associés ua nom de l'espèce correspondante.

In [24]:
Dcoords = {'iris setosa': [], 'iris virginica': [], 'iris versicolor': []} # initialisation du dictionnaire
## 
Despeces = {0: 'iris setosa',
            1: 'iris virginica',
            2: 'iris versicolor'} # correspondances entre numéros et noms pour chaque espèce
for t in data:
    longueur, largeur, numero_espece = t
    Dcoords[Despeces[numero_espece]].append((longueur, largeur))

On obtient le dictionnaire suivant :

In [25]:
Dcoords

{'iris setosa': [(1.4, 0.2), (1.4, 0.2), (1.3, 0.2), (1.5, 0.2), (1.4, 0.2), (1.7, 0.4), (1.4, 0.3), (1.5, 0.2), (1.4, 0.2), (1.5, 0.1), (1.5, 0.2), (1.6, 0.2), (1.4, 0.1), (1.1, 0.1), (1.2, 0.2), (1.5, 0.4), (1.3, 0.4), (1.4, 0.3), (1.7, 0.3), (1.5, 0.3), (1.7, 0.2), (1.5, 0.4), (1.0, 0.2), (1.7, 0.5), (1.9, 0.2), (1.6, 0.2), (1.6, 0.4), (1.5, 0.2), (1.4, 0.2), (1.6, 0.2), (1.6, 0.2), (1.5, 0.4), (1.5, 0.1), (1.4, 0.2), (1.5, 0.1), (1.2, 0.2), (1.3, 0.2), (1.5, 0.1), (1.3, 0.2), (1.5, 0.2), (1.3, 0.3), (1.3, 0.3), (1.3, 0.2), (1.6, 0.6), (1.9, 0.4), (1.4, 0.3), (1.6, 0.2), (1.4, 0.2), (1.5, 0.2), (1.4, 0.2)], 'iris virginica': [(4.7, 1.4), (4.5, 1.5), (4.9, 1.5), (4.0, 1.3), (4.6, 1.5), (4.5, 1.3), (4.7, 1.6), (3.3, 1.0), (4.6, 1.3), (3.9, 1.4), (3.5, 1.0), (4.2, 1.5), (4.0, 1.0), (4.7, 1.4), (3.6, 1.3), (4.4, 1.4), (4.5, 1.5), (4.1, 1.0), (4.5, 1.5), (3.9, 1.1), (4.8, 1.8), (4.0, 1.3), (4.9, 1.5), (4.7, 1.2), (4.3, 1.3), (4.4, 1.4), (4.8, 1.4), (5.0, 1.7), (4.5, 1.5), (3.5, 1.0), (3.

* représentation graphique

Pour construire la représentation graphique demandée, on pourra utiliser un dictionnaire des couleurs à utiliser :

In [26]:
Dcouleurs = {'iris setosa': 'red', 'iris virginica': 'blue', 'iris versicolor': 'green'}

et on devra, pour chaque espèces construire une liste des abscisses pour chaque fleur (liste de longueurs de pétales), et la liste des ordonnées correspondantes (liste de largeurs de pétales), afin de pouvoir afficher les points correspondants.  
   
On pourra là encore utilement recourir à un dictionnaire :

In [27]:
DLxLy = {'iris setosa': ([], []), 'iris virginica': ([], []), 'iris versicolor': ([], [])}
for espece, coords in Dcoords.items():
    for coord in coords:
        x, y = coord
        DLxLy[espece][0].append(x) # ajout de l'abscisse de la fleur à la liste des abscisses pour l'espèce traitée
        DLxLy[espece][1].append(y) # ajout de l'ordonnée de la fleur à la liste des abscisses pour l'espèce traitée

On obtient un dictionnaire dont les clés sont les noms des trois espèces avec pour valeur associée un couple de listes, dont la première est la liste des abscisses pour les fleurs de cette espèce dans l'échantillon, et la seconde la liste correspondante de leurs ordonnées :

In [28]:
DLxLy

{'iris setosa': ([1.4, 1.4, 1.3, 1.5, 1.4, 1.7, 1.4, 1.5, 1.4, 1.5, 1.5, 1.6, 1.4, 1.1, 1.2, 1.5, 1.3, 1.4, 1.7, 1.5, 1.7, 1.5, 1.0, 1.7, 1.9, 1.6, 1.6, 1.5, 1.4, 1.6, 1.6, 1.5, 1.5, 1.4, 1.5, 1.2, 1.3, 1.5, 1.3, 1.5, 1.3, 1.3, 1.3, 1.6, 1.9, 1.4, 1.6, 1.4, 1.5, 1.4], [0.2, 0.2, 0.2, 0.2, 0.2, 0.4, 0.3, 0.2, 0.2, 0.1, 0.2, 0.2, 0.1, 0.1, 0.2, 0.4, 0.4, 0.3, 0.3, 0.3, 0.2, 0.4, 0.2, 0.5, 0.2, 0.2, 0.4, 0.2, 0.2, 0.2, 0.2, 0.4, 0.1, 0.2, 0.1, 0.2, 0.2, 0.1, 0.2, 0.2, 0.3, 0.3, 0.2, 0.6, 0.4, 0.3, 0.2, 0.2, 0.2, 0.2]), 'iris virginica': ([4.7, 4.5, 4.9, 4.0, 4.6, 4.5, 4.7, 3.3, 4.6, 3.9, 3.5, 4.2, 4.0, 4.7, 3.6, 4.4, 4.5, 4.1, 4.5, 3.9, 4.8, 4.0, 4.9, 4.7, 4.3, 4.4, 4.8, 5.0, 4.5, 3.5, 3.8, 3.7, 3.9, 5.1, 4.5, 4.5, 4.7, 4.4, 4.1, 4.0, 4.4, 4.6, 4.0, 3.3, 4.2, 4.2, 4.2, 4.3, 3.0, 4.1], [1.4, 1.5, 1.5, 1.3, 1.5, 1.3, 1.6, 1.0, 1.3, 1.4, 1.0, 1.5, 1.0, 1.4, 1.3, 1.4, 1.5, 1.0, 1.5, 1.1, 1.8, 1.3, 1.5, 1.2, 1.3, 1.4, 1.4, 1.7, 1.5, 1.0, 1.1, 1.0, 1.2, 1.6, 1.5, 1.6, 1.5, 1.3, 1.3, 1.3, 1.2, 1

* représentation des nuages de points

Pour chaque espèce, on construit le nuage de points correspondant, avec la couleur désirée.

In [29]:
import matplotlib.pyplot as plt

fig, ax = plt.subplots() # nécessaire pour des affichages corrects dans Capytale

for espece in DLxLy.keys():
    Lx, Ly = DLxLy[espece]
    plt.plot(Lx, Ly, marker = 'o', linestyle = '', color = Dcouleurs[espece],
             label = espece # on ajoute une étiquette pour légender la figure
            )
plt.legend() ## pour afficher les étiquettes des nuages de points
plt.show()

### 1.3. Mise en place de la méthode des $k$-plus proches voisins ($k$-nearest neighbors, knn)

**Q4.**

Entre deux fleurs dont les coordonnées, (longueur, largeur), sont données par les couples $(x, y)$ et $(x', y')$, la distance est donnée par le calculde la distance euclidienne :  
$$\Delta =  \sqrt{\left((x'-x)^2+(y'-y)^2\right)}.$$

Si les coordonnées sont transmises à la fonction sous la forme de deux tuples, la fonction peut s'implémenter comme suit :

In [30]:
from math import sqrt

def distance(t1, t2):
    return sqrt((t2[0] - t1[0]) ** 2 + (t2[1] - t1[1]) ** 2)

* test

In [31]:
distance((1, 0), (0, 1))

1.4142135623730951

**Q5.**

Une fleur à classifier étant décrite par un couple $t=(longueur, largeur)$ étant donnée, on veut renvoyer une liste dont les éléments sont les couples, (distance, espece), dont chacun donne la distance de la fleur à classifier à une fleur de l'échantillon (première composante du couple) et l'espèce à laquelle appartient la fleur de l'échantillon utilisée pour le calcul de distance (deuxième composante du couple).

In [32]:
def liste_distances(data, t):
    x0, y0 = t ## longueur, largeur pour la fleur à classer
    L = []
    for fleur in data:
        x, y, espece = fleur #longueur, largeur, espece pour cette fleur de l'échantillon
        ## ou bien :
        ## x, y = fleur[0], fleur[1]
        dist = distance((x0, y0), (x, y))
        L.append((dist, espece))
    return L

* test

In [33]:
liste_distances(data, (2.5, 1.0)) # pour une fleur dont la longueur de pétale est 2.5 et la largeur 1.0

[(1.3601470508735445, 0), (1.3601470508735445, 0), (1.4422205101855958, 0), (1.2806248474865698, 0), (1.3601470508735445, 0), (1.0, 0), (1.3038404810405297, 0), (1.2806248474865698, 0), (1.3601470508735445, 0), (1.3453624047073711, 0), (1.2806248474865698, 0), (1.2041594578792296, 0), (1.4212670403551897, 0), (1.6643316977093237, 0), (1.5264337522473748, 0), (1.16619037896906, 0), (1.3416407864998738, 0), (1.3038404810405297, 0), (1.063014581273465, 0), (1.2206555615733703, 0), (1.1313708498984762, 0), (1.16619037896906, 0), (1.7, 0), (0.9433981132056605, 0), (1.0, 0), (1.2041594578792296, 0), (1.0816653826391966, 0), (1.2806248474865698, 0), (1.3601470508735445, 0), (1.2041594578792296, 0), (1.2041594578792296, 0), (1.16619037896906, 0), (1.3453624047073711, 0), (1.3601470508735445, 0), (1.3453624047073711, 0), (1.5264337522473748, 0), (1.4422205101855958, 0), (1.3453624047073711, 0), (1.4422205101855958, 0), (1.2806248474865698, 0), (1.3892443989449805, 0), (1.3892443989449805, 0), (

**Q6.**

On rappelle une implémentation du tri-insertion en place pour une liste de valeurs comparables entre elles par `<=` :

In [34]:
def tri_insertion(L):
    for i in range(1, len(L)):
        # insertion de la valeur L[i] (recopiée dans tmp) dans la liste déjà triée L[:i]
        tmp = L[i]
        j = i ## j désignera la position à laquelle recopier tmp
        while j > 0 and L[j - 1] > tmp:
            L[j] = L[j - 1] ## décalage d'un rang à droite de toutes les valeurs de la partie triée qui sont > tmp
            j = j - 1 # on décremente j pour examiner la valeur suivante (à gauche de la précédente)
        L[j] = tmp

* tests

In [35]:
from random import randint

L = [randint(1, 50) for i in range(50)]
print(L)
tri_insertion(L)
print(L)

[37, 42, 33, 31, 21, 1, 9, 36, 7, 22, 3, 33, 8, 20, 26, 8, 27, 44, 6, 28, 41, 22, 22, 16, 10, 38, 50, 28, 31, 28, 12, 31, 14, 39, 10, 7, 17, 17, 47, 28, 3, 36, 32, 22, 25, 39, 12, 30, 15, 34]
[1, 3, 3, 6, 7, 7, 8, 8, 9, 10, 10, 12, 12, 14, 15, 16, 17, 17, 20, 21, 22, 22, 22, 22, 25, 26, 27, 28, 28, 28, 28, 30, 31, 31, 31, 32, 33, 33, 34, 36, 36, 37, 38, 39, 39, 41, 42, 44, 47, 50]


**Il s'agit ici d'adapter l'implémentation du tri-insertion en place** au fait que la liste est une liste de couples, (distance, espece), que l'on doit trier **dans l'ordre croissant des valeurs de distance**.

In [36]:
def trier(L):
    for i in range(1, len(L)):
        tmp = L[i]
        j = i
        while j > 0 and L[j - 1][0] > tmp[0]: ## ce sont uniquement les valeurs de la première composante que l'on compare
            L[j] = L[j - 1]
            j = j - 1 
        L[j] = tmp

La complexité du tri est celle du tri-insertion, en $O(n^2)$ où $n={\rm len}(L)$, dans le pire cas - liste triée initialement dans l'ordre décroissant ($O(n)$ dans le meilleur cas - liste déjà triée).

* test

In [37]:
L = liste_distances(data, (2.5, 1.0)) # exemple de la question Q5. 
trier(L)
L

[(0.5099019513592785, 1), (0.7999999999999998, 1), (0.7999999999999998, 1), (0.8485281374238571, 0), (0.9433981132056605, 0), (0.9848857801796104, 0), (1.0, 0), (1.0, 0), (1.0, 1), (1.0, 1), (1.063014581273465, 0), (1.0816653826391966, 0), (1.1313708498984762, 0), (1.140175425099138, 1), (1.16619037896906, 0), (1.16619037896906, 0), (1.16619037896906, 0), (1.2000000000000002, 1), (1.2041594578792296, 0), (1.2041594578792296, 0), (1.2041594578792296, 0), (1.2041594578792296, 0), (1.2041594578792296, 0), (1.2206555615733703, 0), (1.2806248474865698, 0), (1.2806248474865698, 0), (1.2806248474865698, 0), (1.2806248474865698, 0), (1.2806248474865698, 0), (1.2806248474865698, 0), (1.3038404810405295, 1), (1.3038404810405297, 0), (1.3038404810405297, 0), (1.3038404810405297, 0), (1.3416407864998738, 0), (1.3453624047073711, 0), (1.3453624047073711, 0), (1.3453624047073711, 0), (1.3453624047073711, 0), (1.3601470508735445, 0), (1.3601470508735445, 0), (1.3601470508735445, 0), (1.36014705087354

Note : en utilisant, si cela était autorisé, la méthode `.sort`, le tri demandé pourrait être implémenté comme suit

In [38]:
L = liste_distances(data, (2.5, 1.0)) # exemple de la question Q5. 
L.sort(key = lambda x: x[0])
L

[(0.5099019513592785, 1), (0.7999999999999998, 1), (0.7999999999999998, 1), (0.8485281374238571, 0), (0.9433981132056605, 0), (0.9848857801796104, 0), (1.0, 0), (1.0, 0), (1.0, 1), (1.0, 1), (1.063014581273465, 0), (1.0816653826391966, 0), (1.1313708498984762, 0), (1.140175425099138, 1), (1.16619037896906, 0), (1.16619037896906, 0), (1.16619037896906, 0), (1.2000000000000002, 1), (1.2041594578792296, 0), (1.2041594578792296, 0), (1.2041594578792296, 0), (1.2041594578792296, 0), (1.2041594578792296, 0), (1.2206555615733703, 0), (1.2806248474865698, 0), (1.2806248474865698, 0), (1.2806248474865698, 0), (1.2806248474865698, 0), (1.2806248474865698, 0), (1.2806248474865698, 0), (1.3038404810405295, 1), (1.3038404810405297, 0), (1.3038404810405297, 0), (1.3038404810405297, 0), (1.3416407864998738, 0), (1.3453624047073711, 0), (1.3453624047073711, 0), (1.3453624047073711, 0), (1.3453624047073711, 0), (1.3601470508735445, 0), (1.3601470508735445, 0), (1.3601470508735445, 0), (1.36014705087354

et si la fonction `sorted` était autorisée (qui ne trie pas en place mais crée une nouvelle liste, triée) :

In [39]:
L = liste_distances(data, (2.5, 1.0)) # exemple de la question Q5. 
L = sorted(L, key = lambda x: x[0])
L

[(0.5099019513592785, 1), (0.7999999999999998, 1), (0.7999999999999998, 1), (0.8485281374238571, 0), (0.9433981132056605, 0), (0.9848857801796104, 0), (1.0, 0), (1.0, 0), (1.0, 1), (1.0, 1), (1.063014581273465, 0), (1.0816653826391966, 0), (1.1313708498984762, 0), (1.140175425099138, 1), (1.16619037896906, 0), (1.16619037896906, 0), (1.16619037896906, 0), (1.2000000000000002, 1), (1.2041594578792296, 0), (1.2041594578792296, 0), (1.2041594578792296, 0), (1.2041594578792296, 0), (1.2041594578792296, 0), (1.2206555615733703, 0), (1.2806248474865698, 0), (1.2806248474865698, 0), (1.2806248474865698, 0), (1.2806248474865698, 0), (1.2806248474865698, 0), (1.2806248474865698, 0), (1.3038404810405295, 1), (1.3038404810405297, 0), (1.3038404810405297, 0), (1.3038404810405297, 0), (1.3416407864998738, 0), (1.3453624047073711, 0), (1.3453624047073711, 0), (1.3453624047073711, 0), (1.3453624047073711, 0), (1.3601470508735445, 0), (1.3601470508735445, 0), (1.3601470508735445, 0), (1.36014705087354

### 1.4. Algorithme $knn$

**Q7.**

Il s'agit d'examiner les $k$ plus proches voisins d'une fleur à classifier (dont on connaît longueur et largeur des pétales).

Sur l'exemple précédent, si on prend $k=5$, les $k$ plus proches voisins sont donnés par :

In [40]:
L = liste_distances(data, (2.5, 1.0)) # exemple de la question Q5.
trier(L)
L[:5]

[(0.5099019513592785, 1), (0.7999999999999998, 1), (0.7999999999999998, 1), (0.8485281374238571, 0), (0.9433981132056605, 0)]

L'espèce majoritaire parmi les cinq fleurs les plus proches de la fleur à classifier est l'espèce 0 (iris setosa).

Les espèces étant étiquetées par les entiers 0, 1 et 2, on peut utiliser une liste à trois éléments pour implémenter les compteurs des trois espèces, ce qui amène à l'implémentation suivante :

In [41]:
def classification_knn(data, t, k):
    L = liste_distances(data, t)
    trier(L)
    compteurs = [0, 0, 0 ] # une liste de trois compteurs compteurs[i] comptant les fleurs de l'espèce i
    for i in range(k): # pour les k plus petites distances obtenues
        espece = L[i][1] # on note l'espèce de la fleur à cette distance
        compteurs[espece] += 1 # on incrémente de 1 le compteur correspondant
    ## recherche du maximum
    nbmax, jmax = compteurs[0], 0 # on initialise avec le nombre de fleurs de l'espèce zéro parmi les k plus proches
    for j in range(3): ## on examine les compteurs des espèces 1 et 2
        if compteurs[j] > nbmax:
            nbmax, jmax = compteurs[j], j
    return jmax #no de l'espèce majoritaire parmi les k plus proches

* tests

In [42]:
classification_knn(data, (2.5, 1.0), 3) ## avec la fleur test de la question Q5. et k = 3

1

In [43]:
classification_knn(data, (2.5, 1.0), 10) ## avec la fleur test de la question Q5. et k = 10

0

On peut aussi utiliser un dictionnaire pour implémenter les compteurs des trois espèces (ce qui permettrait d'utiliser les noms des espèces plutôt que de les numéroter à partir de zéro) :

In [44]:
def classification_knn1(data, t, k):
    L = liste_distances(data, t)
    trier(L)
    # un dictionnaire de compteurs Dcompteurs[espece] comptant les fleurs de l'espèce espece
    Dcompteurs = {0: 0, 1: 0, 2: 0}
    for i in range(k): # pour les k plus petites distances obtenues
        espece = L[i][1] # on note l'espèce de la fleur à cette distance
        Dcompteurs[espece] += 1 # on incrémente de 1 le compteur correspondant
    ## recherche du maximum
    nbmax, emax = Dcompteurs[0], 0 ## on initialise avec le nombre de fleurs de l'une des espèces
                                   ## parmi les k plus proches
    for esp in Dcompteurs.keys(): ## on parcourt le dictionnaire des compteurs par les clés (qui sont les espèces)
        if Dcompteurs[esp] > nbmax:
            nbmax, emax = Dcompteurs[esp], esp
    return emax #espèce majoritaire parmi les k plus proches

* tests

In [45]:
classification_knn1(data, (2.5, 1.0), 3) ## avec la fleur test de la question Q5. et k = 3

1

In [46]:
classification_knn1(data, (2.5, 1.0), 10) ## avec la fleur test de la question Q5. et k = 10

0

* tests

In [47]:
classification_knn(data, (2.5, 1.0), 3) ## avec la fleur test de la question Q5. et k = 3

1

In [48]:
classification_knn(data, (2.5, 1.0), 10) ## avec la fleur test de la question Q5. et k = 10

0

* visualisation 

On représente par un point noir la fleur à classifier, et on trace deux cercles pour pourvoir dénombrer les trois fleurs les plus proches et les dix fleurs les plus proches et leurs espèces.

On fera attention au fait que plusieurs fleurs de l'échantillon ont même longueur et largeur de pétales, et donc certains points figurés correspondent à plusieurs fleurs dans l'échantillon

In [49]:
import matplotlib.pyplot as plt

x, y = (2.5, 1.0) ## longueur et largeur pour la fleur à classifier
# calcul des distances à la fleur à classifier
L = liste_distances(data, (x, y))
trier(L)
# distance et espece de la troisième fleur la plus proche
k = 3
d3, e3 = L[k - 1]
# distance et espece de la dixième fleur la plus proche
k = 10
d10, e10 = L[k - 1]


fig, ax = plt.subplots() # nécessaire pour des affichages corrects dans Capytale

## fleurs de l'échantillon
for triplet in data:
    longueur, largeur, espece = triplet
    if espece == 0:
        plt.plot([longueur], [largeur], 'ro') 
    elif espece == 1:
        plt.plot([longueur], [largeur], 'bo') 
    else:
        plt.plot([longueur], [largeur], 'go') 

## fleur à classifier
plt.plot([x], [y], 'ko')

## cercle passant par la troisème fleur la plus proche
from matplotlib.patches import Circle
##https://matplotlib.org/stable/api/_as_gen/matplotlib.patches.Circle.html#matplotlib.patches.Circle
c3 = plt.Circle((x, y), radius = d3, fill=False)
ax.add_patch(c3)

c10 = plt.Circle((x, y), radius = d10, fill=False)
ax.add_patch(c10)

plt.axis('equal') ## pour que le repère à l'écran soit orthonormé
plt.show()

On pourra zoomer à l'aide de la loupe pour grossir la zone d'intérêt.

On peut afficher les dix plus petites distances et les espèces correspondantes :

In [50]:
L[:10]

[(0.5099019513592785, 1), (0.7999999999999998, 1), (0.7999999999999998, 1), (0.8485281374238571, 0), (0.9433981132056605, 0), (0.9848857801796104, 0), (1.0, 0), (1.0, 0), (1.0, 1), (1.0, 1)]

**Q8.**

La complexité d'un appel à la fonction `classification_knn(data, t, k)` en fonction du nombre $n$ de fleurs dans l'échantillon, du nombre $p$ d'espèces présentes, et de $k$, est la somme des complexités suivantes :  
- $O(n)$ pour le calcul de toutes les distances ;  
- $O(n^2)$ pour le tri de la liste des couples (distance, espèce) ;  
- $O(p+k)$ pour le dénombrement des espèces parmi les $k$ plus proches ($O(p)$ pour la construction de la liste des compteurs (liste de $p$ zéros) et $O(p)$ pour l'examen de l'espèce des $k$ plus proches) ;
- $O(p)$ pour la recherche de l'espèce majoritaire parmi les $k$ plus proches.  
  
La complexité totale est donc en $O(n)+O(n^2)+O(p+k)+O(p)=O(n^2+p+k)$.  
Le nombre d'espèces étant constant, la complexité sera en $O(n^2+k)$.

**Q9.**

La complexité peut être améliorée par le fait qu'il est inutile de trier toute la liste $L$ des couples (distance, espèces).  
On peut se contenter de parcourir la liste $L$ en construisant, au fil du parcours, la liste triée $Lk$ des $k$ plus proches voisins.  
La liste $Lk$ est initialement vide, puis, au fil du parcours de $L$ avec une boucle `for` d'indice $i$ :
- si $Lk$ comporte moins de $k$ éléments, on ajoute $L[i]=(d_i, e_i)$ en dernière position dans $Lk$ ;
- si $Lk$ comporte $k$ éléments, on teste si $L[i][0]$ (distance à la fleur à classifier) est inférieure strictement à la distance $Lk[-1][0]$ et si tel est le cas, on remplace $Lk[-1]$ par $L[i]$ ;
- on déplace le dernier couple dans la liste $Lk$ jusqu'à ce que $Lk$ soit triée (selon le principe du tri insertion). 
  
On suppose (bien sûr) que $k$ est inférieur strictement au nombre d'éléments dans la base.

In [51]:
def classification_knn2(data, t, k):
    L = liste_distances(data, t)
    Lk = []
    for i in range(len(L)):
        if len(Lk) < k:
            Lk.append(L[i])
        else:
            if L[i][0] < Lk[-1][0]:
                Lk[-1] = L[i]
        # décalage de la dernière valeur dans Lk pour maintenir Lk triée
        tmp = Lk[len(Lk) - 1]
        j = len(Lk) - 1
        while j > 0 and Lk[j - 1][0] > tmp[0]: ## ce sont uniquement les valeurs de la première composante que l'on compare
            Lk[j] = Lk[j - 1]
            j = j - 1 
        Lk[j] = tmp
    # à ce stade Lk est la liste des couples (distance, espèce) pour les k plus proches
    
    # un dictionnaire de compteurs Dcompteurs[espece] comptant les fleurs de l'espèce espece
    Dcompteurs = {0: 0, 1: 0, 2: 0}
    for i in range(k): # pour les k plus petites distances obtenues
        espece = Lk[i][1] # on note l'espèce de la fleur à cette distance
        Dcompteurs[espece] += 1 # on incrémente de 1 le compteur correspondant
    ## recherche du maximum
    nbmax, emax = Dcompteurs[0], 0 ## on initialise avec le nombre de fleurs de l'une des espèces
                                   ## parmi les k plus proches
    for esp in Dcompteurs.keys(): ## on parcourt le dictionnaire des compteurs par les clés (qui sont les espèces)
        if Dcompteurs[esp] > nbmax:
            nbmax, emax = Dcompteurs[esp], esp
    return emax #espèce majoritaire parmi les k plus proches

* tests

In [52]:
classification_knn2(data, (2.5, 1.0), 3) ## avec la fleur test de la question Q5. et k = 3

1

In [53]:
classification_knn2(data, (2.5, 1.0), 10) ## avec la fleur test de la question Q5. et k = 10

0

Avec cette modification, la complexité des opérations de tri correspond $n$ insertions dans une liste, $Lk$, comportant au plus $k$ éléments, donc avec une complexité totale (dans le pire cas) en $O(n\times k)$, au lieu de $O(n^2)$.  
La complexité totale de `classification_knn2(data, t, k)` est donc désormais en $O(n)+O(nk)+O(p+k)+O(p)=O(nk+p)$.  
Le nombre d'espèces étant constant, la complexité sera en $O(nk)$.

**Q10.**

Il s'agit pour tout couple $(i\times h, j\times h)$ tel que $0\leqslant ih < 7$ et $0\leqslant jh < 2,5$, de déterminer l'espèce renvoyée par l'appel `classification_knn2(data, (i * h, j * h), k)` et d'afficher un point de la couleur correspondant à l'espèce.

In [54]:
import matplotlib.pyplot as plt

k = 3
h = 0.05
Dcouleurs = {0: 'red', 1: 'blue', 2: 'green'}

fig, ax = plt.subplots()

i = 0
while i * h < 7:
    j = 0
    while j * h < 2.5:
        espece = classification_knn2(data, (i * h, j * h), k)
        plt.fill([i * h, (i + 1) * h, (i + 1) * h, i * h, i * h],
                 [j * h, j * h, (j + 1) * h, (j + 1) * h, j * h],
                 color = Dcouleurs[espece])
        j += 1
    i += 1
plt.show()

In [55]:
from random import sample
help(sample)

Help on method sample in module random:

sample(population, k, *, counts=None) method of random.Random instance
    Chooses k unique random elements from a population sequence or set.
    
    Returns a new list containing elements from the population while
    leaving the original population unchanged.  The resulting list is
    in selection order so that all sub-slices will also be valid random
    samples.  This allows raffle winners (the sample) to be partitioned
    into grand prize and second place winners (the subslices).
    
    Members of the population need not be hashable or unique.  If the
    population contains repeats, then each occurrence is a possible
    selection in the sample.
    
    Repeated elements can be specified one at a time or with the optional
    counts parameter.  For example:
    
        sample(['red', 'blue'], counts=[4, 2], k=5)
    
    is equivalent to:
    
        sample(['red', 'red', 'red', 'red', 'blue', 'blue'], k=5)
    
    To choose a sa

### 1.6 Qualité des prédictions 

**Q11.**

On peut se donner une liste de trois listes vides et y accumuler les éléments de la liste `data` qui ne vérifierait pas l'hypothèse :

In [56]:
L = [[], [], []]
for i in range(3):
    for j in range(i * 50, (i + 1) * 50):
        if data[j][-1] != i: # test de l'espèce pour la fleur d'indice j compris entre 0 et 49, 50 et 99, 100 et 149
            L[i].append((j, data[j]))
print(L)

[[], [], []]


On obtient trois listes vides, ce qui valide l'hypothèse.

**Q12.**

On tire au hasard 5 indices dans chaque tranche d'indices [0, 49], [50, 99] et [100, 149]

In [57]:
from random import sample
I1 = sample([i for i in range(0, 50)], 5)
I2 = sample([i for i in range(50, 100)], 5)
I3 = sample([i for i in range(100, 150)], 5)
I1, I2, I3

([24, 6, 45, 19, 41], [54, 51, 79, 74, 52], [135, 103, 132, 134, 142])

On construit la liste `T` 

In [58]:
T = [data[i] for i in I1 + I2 + I3]

In [59]:
T

[(1.9, 0.2, 0), (1.4, 0.3, 0), (1.4, 0.3, 0), (1.5, 0.3, 0), (1.3, 0.3, 0), (4.6, 1.5, 1), (4.5, 1.5, 1), (3.5, 1.0, 1), (4.3, 1.3, 1), (4.9, 1.5, 1), (6.1, 2.3, 2), (5.6, 1.8, 2), (5.6, 2.2, 2), (5.6, 1.4, 2), (5.1, 1.9, 2)]

On construit la liste `R` des autres fleurs :

In [60]:
I = I1 + I2 + I3
R = []
for i in range(len(data)):
    if i not in I:
        R.append(data[i])

In [61]:
R

[(1.4, 0.2, 0), (1.4, 0.2, 0), (1.3, 0.2, 0), (1.5, 0.2, 0), (1.4, 0.2, 0), (1.7, 0.4, 0), (1.5, 0.2, 0), (1.4, 0.2, 0), (1.5, 0.1, 0), (1.5, 0.2, 0), (1.6, 0.2, 0), (1.4, 0.1, 0), (1.1, 0.1, 0), (1.2, 0.2, 0), (1.5, 0.4, 0), (1.3, 0.4, 0), (1.4, 0.3, 0), (1.7, 0.3, 0), (1.7, 0.2, 0), (1.5, 0.4, 0), (1.0, 0.2, 0), (1.7, 0.5, 0), (1.6, 0.2, 0), (1.6, 0.4, 0), (1.5, 0.2, 0), (1.4, 0.2, 0), (1.6, 0.2, 0), (1.6, 0.2, 0), (1.5, 0.4, 0), (1.5, 0.1, 0), (1.4, 0.2, 0), (1.5, 0.1, 0), (1.2, 0.2, 0), (1.3, 0.2, 0), (1.5, 0.1, 0), (1.3, 0.2, 0), (1.5, 0.2, 0), (1.3, 0.3, 0), (1.3, 0.2, 0), (1.6, 0.6, 0), (1.9, 0.4, 0), (1.6, 0.2, 0), (1.4, 0.2, 0), (1.5, 0.2, 0), (1.4, 0.2, 0), (4.7, 1.4, 1), (4.0, 1.3, 1), (4.5, 1.3, 1), (4.7, 1.6, 1), (3.3, 1.0, 1), (4.6, 1.3, 1), (3.9, 1.4, 1), (3.5, 1.0, 1), (4.2, 1.5, 1), (4.0, 1.0, 1), (4.7, 1.4, 1), (3.6, 1.3, 1), (4.4, 1.4, 1), (4.5, 1.5, 1), (4.1, 1.0, 1), (4.5, 1.5, 1), (3.9, 1.1, 1), (4.8, 1.8, 1), (4.0, 1.3, 1), (4.9, 1.5, 1), (4.7, 1.2, 1), (4.4, 1.4

**Q14.**

On initialise la matrice $C$ :

In [62]:
p = 3 # le nombre d'espèces différentes
C = [[0] * p for i in range(p)]

Puis, pour chaque fleur de l'échantillon, on note son espèce, $i$, et on fait appel à la fonction `classification_knn3` pour obtenir son espèce prédite, $j$, en appliquant l'algorithme des $k$ plus proches voisins pour $k=3$ : on incrémente alors de $1$ le coefficient $C[i][j]$.

In [63]:
k = 3
for fleur in T:
    longueur, largeur, espece = fleur
    espece_predite = classification_knn2(R, (longueur, largeur), k)
    C[espece][espece_predite] += 1

In [64]:
C

[[5, 0, 0], [0, 5, 0], [0, 0, 5]]

On observe (cela peut varier car les ensembles Ret T ont été générés aléatoirement) que l'on obtient (le plus souvent), une matrice diagonale, donc des prédictions exactes quelle que soit l'espèce de la fleur testée.

On peut refaire des tests en changeant la taille $r$ de l'échantillon de référence $R$.  
Précédemment, on a pris $r=135$, le nombre de fleurs, $m$, por chaque espèce dans l'échantillon de test était $m=5$.
On pourra ainsi modifier la valeur de $m$, et la taille de l'échantillon de référence sera $r=150-3m$.  
On pourra aussi faire varier la valeur de $k$.

In [65]:
m = 10 ## échantillon de référence de cardinal 120
k = 3

## création des ensembles T et R
from random import sample

I1 = sample([i for i in range(0, 50)], m)
I2 = sample([i for i in range(50, 100)], m)
I3 = sample([i for i in range(100, 150)], m)

T = [data[i] for i in I1 + I2 + I3]
I = I1 + I2 + I3
R = []
for i in range(len(data)):
    if i not in I:
        R.append(data[i])

## calcul de la matrice de confusion
p = 3 # le nombre d'espèces différentes
C = [[0] * p for i in range(p)]
k = 3
for fleur in T:
    longueur, largeur, espece = fleur
    espece_predite = classification_knn2(R, (longueur, largeur), k)
    C[espece][espece_predite] += 1
print(C)

[[10, 0, 0], [0, 9, 1], [0, 1, 9]]
